services/api/src/models/group.ts

```diff
@@ -922,25 +842,72 @@ export const Group = (clients: {
+          await purgeGroupInCache(members[u].roleSubgroupId)
         } catch (err) {
           throw new Error(`Could not remove user from group: ${err.message}`);
         }
       }
     }
 
-    // once the users are removed from the group, update the cache
-    const allGroups = await loadAllGroups();
-    const keycloakGroups = await transformKeycloakGroups(allGroups);
-    const data = Buffer.from(JSON.stringify(keycloakGroups)).toString('base64')
+    return await loadGroupById(group.id);
+  };
+
+  const purgeGroupInCache =async (groupId: string) => {
+    // lock the redis interactions
+    let lock = await redlock.acquire(["purgelock"], 5000);
     try {
+      let keycloakGroups = []
+      // attempt to retrieve the data from the cache, or fall back to getting from keycloak
+      try {
+        // check redis for the allgroups cache value
+        const data = await getRedisKeycloakCache("allgroups");
+        let buff = new Buffer(data, 'base64');
+        keycloakGroups = JSON.parse(buff.toString('utf-8'));
+      } catch (err) {
+        lock = await lock.extend(5000);
+        logger.warn(`Couldn't check redis keycloak cache: ${err.message}`);
+        // if it can't be recalled from redis, get the data from keycloak
+        const allGroups = await loadAllGroups();
+        keycloakGroups = await transformKeycloakGroups(allGroups);
+        const data = Buffer.from(JSON.stringify(keycloakGroups)).toString('base64')
+        try {
+          // then attempt to save it to redis
+          await saveRedisKeycloakCache("allgroups", data);
+        } catch (err) {
+          logger.warn(`Couldn't save redis keycloak cache: ${err.message}`);
+        }
+      }
+      let fixedKeycloakGroups = keycloakGroups
+      try {
+        // try and replace or update the group within the cache
+        const group = await loadGroupById(groupId);
+        let exists = false;
+        for (const g of keycloakGroups) {
+          if (g.id == groupId){
+            exists = true
+            fixedKeycloakGroups = keycloakGroups.map(item => item.id === groupId ? group : item)
+          }
+        }
+        if (exists === false) {
+          fixedKeycloakGroups.push(group)
+        }
+      } catch (err) {
+        // if the group doesn't exist in keycloak, purge it from the cache
+        fixedKeycloakGroups.splice(fixedKeycloakGroups.findIndex(item => item.id === groupId), 1)
+      }
+      // base64 encode the result
+      const data = Buffer.from(JSON.stringify(fixedKeycloakGroups)).toString('base64')
       // then attempt to save it to redis
-      await saveRedisKeycloakCache("allgroups", data);
-    } catch (err) {
-      logger.warn(`Couldn't save redis keycloak cache: ${err.message}`);
+      try {
+        await saveRedisKeycloakCache("allgroups", data);
+      } catch (err) {
+        logger.warn(`Couldn't save redis keycloak cache: ${err.message}`);
+      }
+    } finally {
+      // release the lock
+      await lock.release();
```

Comment:
Just a general question - are there any other places where we need to worry about locking the redis kc cache?

Reply:
Most other reads/writes are to different keys, and they're not read or written to very often, I think this one makes the most sense mainly because of how special the data in it is